src/main/java/org/protege/owlapi/concurrent/WriteSafeOWLOntologyImpl.java

```diff
@@ -1318,4 +1297,861 @@ public boolean isTopEntity() {
         }
     } 
     
+    @Override
+    public void accept(OWLNamedObjectVisitor visitor) {
+        readLock.lock();
+        try {
+            delegate.accept(visitor);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public <O> O accept(OWLNamedObjectVisitorEx<O> visitor) {
+        readLock.lock();
+        try {
+            return delegate.accept(visitor);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public void setOWLOntologyManager(OWLOntologyManager saveOntology) {
+        writeLock.lock();
+        try {
+            delegate.setOWLOntologyManager(saveOntology);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public void saveOntology() throws OWLOntologyStorageException {
+        writeLock.lock();
```

Comment:
If a read lock is locked then not changes can be applied.  The ReadWriteLock allows for multiple concurrent readers but only a single writer.  The write lock is only for things that change the actual ontology.  Well, that how I think it *should* work!

I think all of this needs starting afresh.  I don't feel confident that ProtegeOWLOntologyManager will work after this. 

Oh, and there's way too many different things going on in this library :(   Might separate out the concurrent stuff.

btw are the streams in v5 streams on copies of collections?